src/main/scala/mesosphere/marathon/api/TaskKiller.scala

```diff
@@ -20,17 +21,27 @@ class TaskKiller @Inject() (
     val authorizer: Authorizer) extends AuthResource {
 
   def kill(appId: PathId,
-           findToKill: (Iterable[Task] => Iterable[Task]))(implicit identity: Identity): Future[Iterable[Task]] = {
+           findToKill: (Iterable[Task] => Iterable[Task]),
+           wipe: Boolean = false)(implicit identity: Identity): Future[Iterable[Task]] = {
     result(groupManager.app(appId)) match {
       case Some(app) =>
         checkAuthorization(UpdateApp, app)
 
-        val tasks = taskTracker.appTasksLaunchedSync(appId)
-        val toKill = findToKill(tasks)
-        if (toKill.nonEmpty)
-          service.killTasks(appId, toKill)
+        import scala.concurrent.ExecutionContext.Implicits.global
+        taskTracker.appTasks(appId).map { allTasks =>
+          val allFound = findToKill(allTasks)
+
+          if (wipe) allFound.foreach { task =>
+            stateOpProcessor.process(TaskStateOp.ForceExpunge(task.taskId))
```

Comment:
you need to wait for the future, i.e. chain the processing step so that you only acknowledge the API call once the task op has been processed.